keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,309 @@
+
+![](/Users/jcope/Workspace/go/src/github.com/yard-turkey/enhancements/keps/sig-storage/bucket-static-api-relation.png)
+
+
+
+### Provisioner Secrets
+
+Per [Non-Goals](#non-goals), it is not within the scope of the proposal to abstract IAM operations.  Instead, provisioner and user credentials should be provided to automation by admins or users.  
+
+BucketClass.Parameter key values are used to specify Secret name and namespace.  These may be considered of different degrees of granularity.
+
+- **Per Provisioner:** the Secret is used for all provisioning operations.  These Secrets should be injected directly in provisioner containers via [common Kubernetes patterns](https://kubernetes.io/docs/tasks/inject-data-application/distribute-credentials-secure/).
+- **Per Operation/Bucket:** Secret’s data is passed to provisioners per operation or Bucket.  These Secrets should be defined as BucketClass parameters.  The keys should be clearly defined during automation design.
+
+This model should support both plain string values as well as “templates.”  Plain string values will allow admins to specify a particular Secret by name and namespace.  The key values should provide a means of defining *provisioner* credentials as well as *user* credentials.  User credentials may be required in cases where a user is being granted access to an existing bucket or when a user has permissions to create buckets themselves.
+
+As an example, the following key’s represent a minimun of what automation should support.  `provisioner-secret*` keys specify Secret to be used for bucket operations.  `user-secret-*` keys specify Secrets containing user identity information.
+
+```yaml
+cosi.io/provisioner-secret-name:
+cosi.io/provisioner-secret-namespace:
+cosi.io/user-secret-name:
+cosi.io/user-secret-namespace:
+```
+
+Bucket automation should support templating of Secret names and namespaces. For example:
+
+```yaml
+cosi.io/user-secret-name: "${bucket.name}"
+cosi.io/user-secret-namespace: "${bucket.namespace}-region-west"
+```
+
+> Note: Annotation key templates are not mentioned as of yet.  However, need for them may arise during automation design.
+
+## Custom Resource Definitions
+
+#### Bucket
+
+A user facing API object representing an object store bucket. Created by a user in their app's namespace. Once provisiong is complete, the `Bucket` is "bound" to the corresponding `BucketContent`. Binding is 1:1, meaning there is only one `BucketContent` per `Bucket` and vice-versa.
```

Comment:
Not being able to use a single bucket (in the storage system) from workloads running in  two (or more) namespaces seems quite restrictive...

Reply:
Just like nfs volume(shared filesystem), pvc object is also namespaced.
If their is necessary to share object storage between namespaces, maybe create 2 bucket-bucketcontent pairs which confiured as the same backend.